sdks/python/apache_beam/runners/common.py

```diff
@@ -495,6 +525,162 @@ def invoke_process(self,
+    return self._watermark_estimator.current_watermark()
+
+  def observe_timestamp(self, timestamp):
+    if not isinstance(timestamp, Timestamp):
+      raise ValueError('Input of observe_timestamp should be a Timestamp '
+                       'object')
+    with self._lock:
+      self._watermark_estimator.observe_timestamp(timestamp)
+
+
+class _ThreadsafeRestrictionTracker(object):
+  """A thread-safe wrapper which wraps a `RestritionTracker`.
+
+  This wrapper guarantees synchronization of modifying restrictions across
+  multi-thread.
+  """
+
+  def __init__(self, restriction_tracker, lock):
+    if not isinstance(restriction_tracker, iobase.RestrictionTracker):
+      raise ValueError(
+          'Initialize ThreadsafeRestrictionTracker requires'
+          'RestrictionTracker.')
+    self._restriction_tracker = restriction_tracker
+    # Records an absolute timestamp when defer_remainder is called.
+    self._deferred_timestamp = None
+    self._lock = lock
+    self._deferred_residual = None
+    self._deferred_watermark = None
+
+  def current_restriction(self):
+    with self._lock:
+      return self._restriction_tracker.current_restriction()
+
+  def try_claim(self, position):
+    with self._lock:
+      return self._restriction_tracker.try_claim(position)
+
+  def defer_remainder(self, deferred_time=None):
+    """Performs self-checkpoint on current processing restriction with an
+    expected resuming time.
+
+    Self-checkpoint could happen during processing elements. When executing an
+    DoFn.process(), you may want to stop processing an element and resuming
+    later if current element has been processed quit a long time or you also
+    want to have some outputs from other elements. ``defer_remainder()`` can be
+    called on per element if needed.
+
+    Args:
+      deferred_time: A relative ``Duration`` that indicates the ideal time gap
+      between now and resuming, or an absolute ``Timestamp`` for resuming
+      execution time. If the time_delay is None, the deferred work will be
+      executed as soon as possible.
+    """
+
+    # Record current time for calculating deferred_time later.
+    self._deferred_timestamp = Timestamp.now()
+    if (deferred_time and
+        not isinstance(deferred_time, Duration) and
+        not isinstance(deferred_time, Timestamp)):
+      raise ValueError('The timestamp of deter_remainder() should be a '
+                       'Duration or a Timestamp, or None.')
+    self._deferred_watermark = deferred_time
+    checkpoint = self.try_split(0)
+    if checkpoint:
+      _, self._deferred_residual = checkpoint
+
+  def check_done(self):
+    with self._lock:
+      return self._restriction_tracker.check_done()
+
+  def current_progress(self):
+    with self._lock:
+      return self._restriction_tracker.current_progress()
+
+  def try_split(self, fraction_of_remainder):
+    # The caller should hold the lock before entering this function.
```

Comment:
```
assert _lock.locked(), 'Expected lock to be held to guarantee thread-safe access.'
```
here and elsewhere instead of the comment.

Reply:
Thanks! Changed it to `raise RuntimeError`